[UXUI-160] DNC search command

andersonjeccel opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 7.x)
* a.b for any bug fixes (e.g. 5.2, 6.0)
* c.x for any bug fixes, features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 7.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ❌ <!-- Use emojis to indicate positive (green) or negative (red) for each item in the table. -->
| New feature/enhancement? (use the a.x branch)      | ✔️
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ❌ <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/user-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation-new#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

## Description

### Summary

A new search command is needed to display contacts flagged as Do Not Contact (DNC).

### Context

Currently, there is no way to filter contacts marked as DNC. This feature will allow users to easily find these contacts.

### Acceptance criteria

- Implement a search command to show contacts flagged as DNC.
- Use 'dnc:any' to filter for any channel.
- Use 'dnc:<channel>' for filtering by a specific channel (e.g., dnc:email, dnc:sms).

### Other information

N/A

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->
<!-- Remove HTML comment markup below to use the table for screenshots when relevant. -->
<!--
| Before                                 | After
| -------------------------------------- | ---
|                                        | 
-->


---
### 📋 Steps to test this PR:

<!--
This part is crucial. Take the time to write very clear, annotated and step by step test instructions, because testers may not be developers.
-->

**Having contacts previously created is needed!**

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Open Contacts
3. Select some > Click on actions and Batch Set Do Not Contact
4. Apply
5. On the search field, type dnc:any
6. See if only the contacts flagged as DNC appear

<!--
If you have any deprecations and backwards compatibility breaks, list them here along with the new alternative.
-->

andersonjeccel committed:
[UXUI-160] DNC search command

andersonjeccel committed:
Merge branch '7.x' into UXUI-160-a-page-to-see-all-contacts-tagged-as-dnc-or-search-command

escopecz commented on app/bundles/LeadBundle/Entity/LeadRepository.php:
```diff
@@ -917,6 +917,35 @@ protected function addSearchCommandWhereClause($q, $filter): array
                 break;
+            case $this->translator->trans('mautic.lead.lead.searchcommand.dnc'):
+            case $this->translator->trans('mautic.lead.lead.searchcommand.dnc', [], null, 'en_US'):
+                $anyKeyword = $this->translator->trans('mautic.lead.lead.searchcommand.dnc.any');
+                if ($string === $anyKeyword) {
```
The way it works now is that we are filtering for the filter keyword in English as well as in the local language. This way the `mautic.lead.lead.searchcommand.dnc.any` translation works only for English. Let's build it the same way as other search commands that allow also for the local language keyword.

escopecz reviewed (changes requested):
Notice the `if` and `else` blocks have several rows that are the same. Those rows should be above the `if` statement instead of duplicating it.

kingsedem reviewed (approved):
I have confirmed this and it works as expected.

**Before**

![image](https://github.com/user-attachments/assets/695c529b-c6d0-41f6-a050-e9f194eefb76)


**After**
![image](https://github.com/user-attachments/assets/614fbd58-adb3-4a0c-8663-8eff861d6831)

kuzmany reviewed (approved):
Technically it works 👍 beside @escopecz  note.

I would probably keep consistency in commands:

is:unpublished
is:mine
is:dnc
is:bounce
is:unsubscribe

etc.

But it's minor. I can live with it.

JonasLudwig1998 committed:
Fixing phpstan and fixcs problem

Fixing problem when functional testing with most recent ddev version

Run fixcs

Use direct string interpolation instead of env variable syntax in DB commands

RCheesley committed:
Merge branch '7.x' into UXUI-160-a-page-to-see-all-contacts-tagged-as-dnc-or-search-command

andersonjeccel committed:
Address review from John

andersonjeccel committed:
Add test coverage

andersonjeccel committed:
Revert "Fixing phpstan and fixcs problem"

This reverts commit efff07e379dd3327fd193d56aa533dd8370af59f.

andersonjeccel committed:
Improve code quality

andersonjeccel committed:
CS fix

andersonjeccel committed:
Merge branch '7.x' into UXUI-160-a-page-to-see-all-contacts-tagged-as-dnc-or-search-command

shinde-rahul commented on app/bundles/LeadBundle/Entity/LeadRepository.php:
```diff
@@ -940,6 +940,31 @@ protected function addSearchCommandWhereClause($q, $filter): array
+                            )
+                        );
+                    $returnParameter = true;
+                }
```
```suggestion
                $sq = $this->getEntityManager()->getConnection()->createQueryBuilder();
                $sq->select('1')
                    ->from(MAUTIC_TABLE_PREFIX . 'lead_donotcontact', 'dnc')
                    ->where($q->expr()->eq('l.id', 'dnc.lead_id'));

                if ($string === $anyKeyword || $string === $anyKeywordEn) {
                    $returnParameter = false;
                } else {
                    $sq->andWhere($q->expr()->eq('dnc.channel', ":$unique"));
                    $returnParameter = true;
                }
```

shinde-rahul reviewed (changes requested):
Looks good overall! 

A suggestion to improve readability and reduce duplication:
Both branches of the condition repeat the same base `QueryBuilder`. You can initialize it once and append `->andWhere()` only when channel match is needed.

andersonjeccel committed:
Update app/bundles/LeadBundle/Entity/LeadRepository.php

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

andersonjeccel committed:
style: apply cs fixer to LeadRepository.php

andersonjeccel committed:
Merge branch '7.x' into UXUI-160-a-page-to-see-all-contacts-tagged-as-dnc-or-search-command

escopecz reviewed (approved):
The code looks fine 👍
